### PR TITLE
Brian/implementation details

### PR DESCRIPTION
**Pull request by Geometer1729**
Adds in some of my work on the implementation details

### CODE REVIEW

**Comment by CSVdB** [comment]
I changed the base to `master`, because `implementation-details` was merged into `master`.

File: docs/whitepaper.tex [hunk 1]
```diff
@@ -220,9 +244,105 @@ \subsection{DEX pool}
...
+\begin{itemize}
+	\item The input and output pools are valid
+	\item The redeemer is valid
+	\item When the redeemer action is Swap:
+		\begin{itemize}
+			\item The exchange invariant doesn't decrease
+			\item The Admin fee is increased appropriately
+			\item The Fee amount is left in the pool
+			\item No liquidity tokens are minted
+			\item The input and output pools are both live
```
[the remark this answers is not in the document]
Comment (CSVdB) [reply docs/whitepaper.tex:271]: The datum should probably contain, for each asset class (A and B), both the non-admin balance and the admin balance. That way, any transaction can check that `non-admin + admin = total (in the value)`.

File: docs/whitepaper.tex [hunk 2]
```diff
@@ -220,9 +241,106 @@ \subsection{DEX pool}
   \item A boolean representing the pool's kill switch
 \end{itemize}
 
+
+\subsection{Pool Addresses validator}
+
+The pool addresses validator controls when pools can be spent.
```
[the remark this answers is not in the document]
Comment (CSVdB) [reply docs/whitepaper.tex:253]: What is the "Pool Addresses Validator"? Are we talking about the `Pool Validator Script`, or the `Pool ID Minting Policy`?

File: docs/whitepaper.tex (hunk 1 above)
[the remark this answers is not in the document]
Comment (Geometer1729) [reply docs/whitepaper.tex:271]: Sounds good :+1:

File: docs/whitepaper.tex (hunk 2 above)
[the remark this answers is not in the document]
Comment (Geometer1729) [reply docs/whitepaper.tex:253]: Pool Validator Script

File: docs/whitepaper.tex [hunk 3]
```diff
@@ -220,9 +247,113 @@ \subsection{DEX pool}
...
+			\item The admin signed the transaction
+			\item The output pool is dead
+			\item Nothing else is changed in the pool
+		\end{itemize}
+	\item When The redeemer action is Collect
+		\begin{itemize}
+			\item The admin fees are changed correctly
+				for the amount of value removed from the pool
+			\item The admin fees remain non-negative
+			\item The admin keys signed the tx
```
[the remark this answers is not in the document]
Comment (Geometer1729) [reply docs/whitepaper.tex:293]: For keys? Doesn't the spec say there are 3 admin keys?

File: docs/whitepaper.tex [hunk 4]
```diff
@@ -220,9 +247,113 @@ \subsection{DEX pool}
...
+			\item The input and output pools are both live
+		\end{itemize}
+	\item When the redeemer action is Liquidity
+		\begin{itemize}
+			\item The liquidity invariant $L^2/k^2$ is non decreasing
+			\item Appropriate fees are paid
+				\todo{how should fees work here?}
+			\item The datum correctly reflects the change in issued liquidity
+			\item The "liveness" of the pool is not changed
+			\item If the input pool is dead the issued liquidity decreases
```
[the remark this answers is not in the document]
Comment (Geometer1729) [reply docs/whitepaper.tex:280]: Fair enough, I'll add that the withdraw is balanced as a constraint.

File: docs/whitepaper.tex (hunk 3 above)
[the remark this answers is not in the document]
Comment (marijanp) [reply docs/whitepaper.tex:293]: It's one of the admin keys, not all of them right?

File: docs/whitepaper.tex [hunk 5]
```diff
@@ -220,9 +244,105 @@ \subsection{DEX pool}
...
+	\item The redeemer is valid
+	\item When the redeemer action is Swap:
+		\begin{itemize}
+			\item The exchange invariant doesn't decrease
+			\item The Admin fee is increased appropriately
+			\item The Fee amount is left in the pool
+			\item No liquidity tokens are minted
+			\item The input and output pools are both live
+		\end{itemize}
+	\item When the redeemer action is Liquidity
```
[the remark this answers is not in the document]
Comment (Geometer1729) [reply docs/whitepaper.tex:273]: How to handle fees here is part of the problems and solutions document I sent you so I'm waiting for you to get back to me before I can clarify how that will work here.

File: docs/whitepaper.tex [hunk 6]
```diff
@@ -220,9 +247,118 @@ \subsection{DEX pool}
...
+To do so it should enforce all of the following:
+
+\begin{itemize}
+	\item The config utxo is valid
+	\item Exactly one token is minted
+	\item That token is sent to a pool
+	\item The redeemer provides a seed utxo
+	\item The token name of the minted token is the hash of the seed utxo
+	\item The seed utxo is spent in the transaction
+	\item The output pool datum is valid
```
Comment (CSVdB) [inline docs/whitepaper.tex:318]: What does `valid` mean exactly, in this context?

File: docs/whitepaper.tex [hunk 7]
```diff
@@ -220,9 +247,118 @@ \subsection{DEX pool}
...
+	\item The datum correctly reflects the number of liquidity tokens being minted
```
Comment (CSVdB) [inline docs/whitepaper.tex:319]: What does `correctly` mean, here? Please combine with the last requirement.

File: docs/whitepaper.tex [hunk 8]
```diff
@@ -220,9 +247,118 @@ \subsection{DEX pool}
...
+	\item at least one liquidity token is being minted
+	\item the initial invariant is not zero ie. some liquidity is provided in each token
+	\item The issued liquidity is exactly the square root of the initial value
```
Comment (CSVdB) [inline docs/whitepaper.tex:322]: What is "the initial value"? Feel free to repeat parts from earlier in the spec here.

Comment (CSVdB) [reply docs/whitepaper.tex:322]: What does "exactly" mean, here? It's very unlikely that the square root of `x_0 \cdot y_0` is an integer.

File: docs/whitepaper.tex [hunk 9]
```diff
@@ -220,9 +247,118 @@ \subsection{DEX pool}
...
+\begin{itemize}
+	\item The config utxo is valid
+	\item Exactly one token is minted
+	\item That token is sent to a pool
+	\item The redeemer provides a seed utxo
+	\item The token name of the minted token is the hash of the seed utxo
+	\item The seed utxo is spent in the transaction
+	\item The output pool datum is valid
+	\item The datum correctly reflects the number of liquidity tokens being minted
+	\item at least one liquidity token is being minted
```
Comment (CSVdB) [inline docs/whitepaper.tex:320]: Nitpick: All bullet points should start with a capital letter.

File: docs/whitepaper.tex [hunk 10]
```diff
@@ -220,9 +247,118 @@ \subsection{DEX pool}
...
+	\item the initial invariant is not zero ie. some liquidity is provided in each token
+	\item The issued liquidity is exactly the square root of the initial value
+\end{itemize}
+
+NOTE: Pools not being duplicates will be checked in the browser app, but not enforced onchain.
```
Comment (CSVdB) [inline docs/whitepaper.tex:325]: Explain what you mean by "duplicates".

File: docs/whitepaper.tex [hunk 11]
```diff
@@ -220,9 +247,118 @@ \subsection{DEX pool}
...
+
+\subsection{Liquidity token minting policy}
+
+The liquidity tokens are used to represent ownership of liquidity in a pool.
+Their minting policy is responsible for ensuring that they are only minted
+during liquidity interactions with a pool and during pool initializations.
+
+Since the Pools Address Validator and the pool id token minting policy
```
Comment (CSVdB) [inline docs/whitepaper.tex:333]: What is the Pools Address Validator?

File: docs/whitepaper.tex [hunk 12]
```diff
@@ -220,9 +247,118 @@ \subsection{DEX pool}
...
+already enforce the logic regarding how much should be minted the liquidity
```
Comment (CSVdB) [inline docs/whitepaper.tex:334]: Comma missing after "minted".

File: docs/whitepaper.tex [hunk 13]
```diff
@@ -220,9 +247,118 @@ \subsection{DEX pool}
...
+token minting policy doesn't have to do much. It's main responsibility is enforcing that
```
Comment (CSVdB) [inline docs/whitepaper.tex:335]: Its*, not It's.

File: docs/whitepaper.tex [hunk 14]
```diff
@@ -220,9 +247,118 @@ \subsection{DEX pool}
...
+some pool is involved either by being opened or an existing pool being used, and that
+the liquidity tokens being created correspond to the pool being used.
+The way we keep track of which liquidity tokens correspond to which pools
+is by having the token name be the id of the pool the liquidity tokens
```
Comment (CSVdB) [inline docs/whitepaper.tex:339]: What do you mean, exactly, by "the ID of the pool"? Which part of the Pool ID NFT is considered here? TokenName, CurrencySymbol..?

File: docs/whitepaper.tex [hunk 15]
```diff
@@ -220,9 +247,118 @@ \subsection{DEX pool}
...
+The liquidity tokens are used to represent ownership of liquidity in a pool.
+Their minting policy is responsible for ensuring that they are only minted
+during liquidity interactions with a pool and during pool initializations.
+
+Since the Pools Address Validator and the pool id token minting policy
+already enforce the logic regarding how much should be minted the liquidity
+token minting policy doesn't have to do much. It's main responsibility is enforcing that
+some pool is involved either by being opened or an existing pool being used, and that
+the liquidity tokens being created correspond to the pool being used.
+The way we keep track of which liquidity tokens correspond to which pools
```
Comment (CSVdB) [inline docs/whitepaper.tex:338]: to which pool*, not pools

File: docs/whitepaper.tex [hunk 16]
```diff
@@ -220,9 +247,118 @@ \subsection{DEX pool}
...
+is by having the token name be the id of the pool the liquidity tokens
+belong to.
+
+To do that it enforces the following:
+\begin{itemize}
+	\item The liquidity tokens being minted all have the same token name/pool id
+	\item The redeemer is valid
+		\begin{itemize}
+			\item When the redeemer is Initialize: A pool id token with the same id is being minted
+			\item When the redeemer is Pool: There is an input with a pool id token with the same id
```
Comment (CSVdB) [inline docs/whitepaper.tex:348]: Do you mean "When the redeemer is Liquidity"?

File: docs/whitepaper.tex [hunk 17]
```diff
@@ -220,9 +247,118 @@ \subsection{DEX pool}
...
+
+Since the Pools Address Validator and the pool id token minting policy
+already enforce the logic regarding how much should be minted the liquidity
+token minting policy doesn't have to do much. It's main responsibility is enforcing that
+some pool is involved either by being opened or an existing pool being used, and that
+the liquidity tokens being created correspond to the pool being used.
+The way we keep track of which liquidity tokens correspond to which pools
+is by having the token name be the id of the pool the liquidity tokens
+belong to.
+
```
Comment (CSVdB) [inline docs/whitepaper.tex:341]: Nitpick: New paragraphs should leave a newline, using `\\`, to make our formatting consistent.

File: docs/whitepaper.tex [hunk 18]
```diff
@@ -220,9 +247,118 @@ \subsection{DEX pool}
...
+		\begin{itemize}
+			\item When the redeemer is Initialize: A pool id token with the same id is being minted
+			\item When the redeemer is Pool: There is an input with a pool id token with the same id
+		\end{itemize}
+\end{itemize}
+
 \subsection{Off-chain notes}
 
-\todo{How to list the pools, and get the pool info?}
+Initialize pool: Initially, $k_0 = x_0 \cdot y_0$ where $x_0$ and $y_0$ are the
```
Comment (CSVdB) [inline docs/whitepaper.tex:354]: `k_0^2`

File: docs/whitepaper.tex [hunk 19]
```diff
@@ -220,9 +247,113 @@ \subsection{DEX pool}
...
+			\item Appropriate fees are paid
+				\todo{how should fees work here?}
+			\item The datum correctly reflects the change in issued liquidity
+			\item The "liveness" of the pool is not changed
+			\item If the input pool is dead the issued liquidity decreases
+		\end{itemize}
+	\item When the redeemer action is Kill
+		\begin{itemize}
+			\item The admin signed the transaction
+			\item The output pool is dead
```
[the remark this answers is not in the document]
Comment (CSVdB) [reply docs/whitepaper.tex:290]: It seems like a good idea to me to enforce that there's only one pool going in and out whenever we're talking about a transaction that "updates" a pool, for the sake of limiting attacks against the protocol.
@epiro-ps thoughts?

File: docs/whitepaper.tex (hunk 4 above)
[the remark this answers is not in the document]
Comment (CSVdB) [reply docs/whitepaper.tex:284]: What are the arguments pro and con making "Withdraw" and "Balanced withdraw" separate on-chain endpoints? (I.e. giving them different redeemers.)

File: docs/whitepaper.tex (hunk 2 above)
[the remark this answers is not in the document]
Comment (epiro-ps) [reply docs/whitepaper.tex:253]: I prefer this just to be called `Pool Validator` when referring to the validator and `Pool Validator Address` when referring to its address.

**Comment by Geometer1729** [comment]
Closing in favor of https://github.com/ArdanaLabs/Danaswap/pull/17

**Pull request closed without merging**